src/core/ext/transport/chaotic_good/server/chaotic_good_server.cc

```diff
@@ -0,0 +1,392 @@
+            GRPC_SLICE_START_PTR(slice.c_slice())));
+        GPR_ASSERT(frame_header.ok());
+        return TrySeq(
+            self->connection_->endpoint_->Read(frame_header->GetFrameLength()),
+            [frame_header = *frame_header,
+             self = self](SliceBuffer buffer) mutable {
+              // Read Setting frame.
+              SettingsFrame frame;
+              // Deserialize frame from read buffer.
+              BufferPair buffer_pair{std::move(buffer), SliceBuffer()};
+              auto status = frame.Deserialize(
+                  &self->connection_->hpack_parser_, frame_header,
+                  absl::BitGenRef(self->connection_->bitgen_),
+                  GetContext<Arena>(), std::move(buffer_pair));
+              GPR_ASSERT(status.ok());
+              bool is_control_endpoint =
+                  std::string(
+                      frame.headers
+                          ->get_pointer(ChaoticGoodConnectionTypeMetadata())
+                          ->Ref()
+                          .as_string_view()) == "control";
+              if (!is_control_endpoint) {
+                // Get connection-id and data-alignment for data endpoint.
+                self->connection_->connection_id_ =
+                    frame.headers
+                        ->get_pointer(ChaoticGoodConnectionIdMetadata())
+                        ->Ref();
+                self->connection_->data_alignment_ =
+                    frame.headers->get(ChaoticGoodDataAlignmentMetadata())
+                        .value();
+              }
+              return is_control_endpoint;
+            });
+      });
+}
+
+auto ChaoticGoodServerListener::ActiveConnection::HandshakingState::
+    WaitForDataEndpointSetup(std::shared_ptr<HandshakingState> self) {
+  return Race(
+      TrySeq(
+          []() {
+            // TODO(ladynana): find a way to resolve SeqState to actual value.
+            return absl::OkStatus();
+          },
+          [self = self]() mutable {
+            MutexLock lock(&self->connection_->listener_->mu_);
+            auto latch =
+                self->connection_->listener_->connectivity_map_
+                    .find(std::string(
+                        self->connection_->connection_id_.as_string_view()))
+                    ->second;
+            return latch->Wait();
+          },
+          [](std::shared_ptr<PromiseEndpoint> ret) -> absl::Status {
+            GPR_ASSERT(ret != nullptr);
+            // TODO(ladynana): initialize server transport.
+            return absl::OkStatus();
+          }),
+      // Set timeout for waiting data endpoint connect.
+      TrySeq(
+          // []() {
+          Sleep(Timestamp::Now() + self->connection_->connection_deadline_),
+          [self = self]() mutable -> absl::Status {
+            MutexLock lock(&self->connection_->listener_->mu_);
+            // Delete connection id from map when timeout;
```

Comment:
Is simply deleting from connectivity_map enough ? Do we need to cancel any pending operations to fail gracefully ?